### ISSUE TITLE
Move some of storage config from command line to config file

### ISSUE THREAD

**Issue by codesome** [prometheus/prometheus#9115]
## Proposal

https://github.com/prometheus/prometheus/pull/8974 introduced a `storage` section in the config file and also added mechanism to reload the config in TSDB without needing a restart. I think we can (and should) move some of the CLI flags to the config file so that it can be reloaded without having to restart Prometheus (hence the WAL replay).

The config file would take precedence over the CLI flags, and deprecate the CLI flags in favour of config file.

Some of the flags that I am proposing to move to file (not including any hidden flags here):

* `--storage.tsdb.retention.time=STORAGE.TSDB.RETENTION.TIME`
* `--storage.tsdb.retention.size=STORAGE.TSDB.RETENTION.SIZE`
* `--storage.tsdb.allow-overlapping-blocks` - this only to allow overlapping blocks if it is disabled. If it's already allowed, config reload should not disable it again. Have to restart Prometheus to disable so that overlapping blocks fail to startup if disabled.

Maybe also
* `--storage.tsdb.wal-compression`

@cstyan do you think we can move some of these below flags to config file too?
```
--storage.remote.flush-deadline=<duration>  
--storage.remote.read-sample-limit=5e7  
--storage.remote.read-concurrent-limit=10  
--storage.remote.read-max-bytes-in-frame=1048576
```

cc @bwplotka

**Comment by roidelapluie** [comment]
Hello,

:+1:

--storage.tsdb.wal-compression is pretty much an internal stuff and users should generally not be aware of that, I would not put it in the config file.

**Comment by codesome** [comment]
> --storage.tsdb.wal-compression is pretty much an internal stuff and users should generally not be aware of that, I would not put it in the config file.

Then I guess we should hide it from `-h` output

**Comment by roidelapluie** [comment]
> > --storage.tsdb.wal-compression is pretty much an internal stuff and users should generally not be aware of that, I would not put it in the config file.
> 
> Then I guess we should hide it from `-h` output

You still need it if you want to keep you WAL backwards compatible with pretty old Prometheus releases, but that's it I think. There was a bug that we fixed recently about very large records: #8790, but that's the only case we have ever needed to disable it, and it is no longer needed since the fix.

Overall yes I think in next releases we could hide the flag.

**Comment by codesome** [comment]
Opened https://github.com/prometheus/prometheus/pull/9117 which should take care of hiding it

**Comment by SuperQ** [comment]
@roidelapluie The TSDB will read both compressed and un-compressed WAL files. The flag is only useful if you wanted to disable compression and roll back to <= 2.19.0.

**Comment by roidelapluie** [comment]
> @roidelapluie The TSDB will read both compressed and un-compressed WAL files. The flag is only useful if you wanted to disable compression and roll back to <= 2.19.0.

Yes that is what I tried to explain, apparently badly :)

**Comment by cstyan** [comment]
Honestly I'd forgotten that `--storage.remote.flush-deadline=<duration> ` was even a CLI flag and not a config file option. TBH it feels to me like that should be a per queue configuration option with the same default as the current CLI flag. @csmarchbanks any thoughts?

For the rest of the `storage.remote` options @bwplotka or someone else with more remote read context should chime in.

**Comment by nicolastakashi** [comment]
Hi peeps!
@codesome I would try to implement this one, can I?

**Comment by codesome** [comment]
@nicolastakashi yes, please go ahead! Thanks.

Since remote flags are not decided yet, can you only do 3 `--storage.tsdb.xx` flags for now for now?

**Comment by nicolastakashi** [comment]
@codesome sure right!

Just to clarify, these flags should be placed on the global config file or should I create a new one?

**Comment by codesome** [comment]
@nicolastakashi Global config file. You will already find a place for `storage`, these flags should be as `storage -> tsdb -> xx` in the yaml.

**Comment by nicolastakashi** [comment]
Nice!
Thank you for the explanation, I'll take a look on that and as soon as possible open a MR

**Comment by csmarchbanks** [comment]
I agree with Callum about flush-deadline, that should be in the config file for sure. The read ones seem reasonable to put in the config file to me, but I would also appreciate someone who uses remote read more than me to chime in.

**Comment by BrutalBirdie** [comment]
We just hit this issue as well.
https://forum.cloudron.io/topic/5633/add-option-for-users-to-add-starting-parameters?_=1631174530839

Seeing this merged would be awesome.

Blocking this PR as well should be changes to the doc https://github.com/prometheus/docs right?

If I can be of assistance let me know.

**Comment by ivange94** [comment]
@codesome what's the update on this? Is someone still working on this if not I'll like to pick this one.

**Comment by codesome** [comment]
@ivange94 There is #9158 open by @nicolastakashi which seems to have gone stale. I have pinged the author on that PR and will wait for any response for a week. If there is no response for a week (or if the author says they are not working on it anymore), you can consider it to be open for you to pick.

**Comment by kg-mpg** [comment]
pinging for status, I'm interested to take a look

**Comment by tashif-hoda** [comment]
@codesome Is someone still working on this? If not, can I take this up?

**Comment by nicolastakashi** [comment]
Hey @tashif-hoda my priorities shift a bit and I wont able to continue this for now.
Feel free to take over it if you wanna.

**Comment by tashif-hoda** [comment]
@codesome @nicolastakashi One clarification regarding "--storage.tsdb.allow-overlapping-blocks". This configuration is now set to True by default and the flag is marked as deprecated and hidden. Do we still need to port this to the config file?

**Comment by bboreham** [comment]
Hello from the bug-scrub!  We have re-asked for a review on #12833.

>  "--storage.tsdb.allow-overlapping-blocks". This configuration is now set to True by default and the flag is marked as deprecated and hidden. Do we still need to port this to the config file?

Consensus on the call was no: this option is quite unlikely to change on-the-fly.

**Comment by baryluk** [comment]
I know this is a bit oldish, but I also do see some movements on this recently.

I am not against config and reload, but I really do like having option to do it via flags. Could we make the flags supported (at least for the major flags for storage) and not deprecated instead?

**Comment by beorn7** [comment]
With a large user base, there is – understandably – a great diversity of preferences how to do a certain thing.

I still think, at the end of the day, it's better to stick with "provide one way to do one thing" philosophy. Otherwise, we'll see an explosion of config options of various kinds that are needlessly confusing.
